# Airline Satisfaction ✈
## Data Cleaning
In this notebook, we aim to prepare the airline customer service survey for a classification machine learning model to be built into a Streamlit app. The goal of this cleanup is to retain and create as many useful numerical columns as possible while addressing null values.

---

## Load Libraries

In [4]:
import pandas as pd

## Load Dataset

In [6]:
df = pd.read_csv('data/train.csv')

In [7]:
# Set the display option to show all columns
pd.set_option('display.max_columns', None)

# Show the first 20 rows
df.head(20)

Unnamed: 0      id  Gender      Customer Type  Age   Type of Travel  \
0            0   70172    Male     Loyal Customer   13  Personal Travel   
1            1    5047    Male  disloyal Customer   25  Business travel   
2            2  110028  Female     Loyal Customer   26  Business travel   
3            3   24026  Female     Loyal Customer   25  Business travel   
4            4  119299    Male     Loyal Customer   61  Business travel   
5            5  111157  Female     Loyal Customer   26  Personal Travel   
6            6   82113    Male     Loyal Customer   47  Personal Travel   
7            7   96462  Female     Loyal Customer   52  Business travel   
8            8   79485  Female     Loyal Customer   41  Business travel   
9            9   65725    Male  disloyal Customer   20  Business travel   
10          10   34991  Female  disloyal Customer   24  Business travel   
11          11   51412  Female     Loyal Customer   12  Personal Travel   
12          12   98628    Male     Loyal Customer   53  Business travel   
13          13   83502    Male     Loyal Customer   33  Personal Travel   
14          14   95789  Female     Loyal Customer   26  Personal Travel   
15          15  100580    Male  disloyal Customer   13  Business travel   
16          16   71142  Female     Loyal Customer   26  Business travel   
17          17  127461    Male     Loyal Customer   41  Business travel   
18          18   70354  Female     Loyal Customer   45  Business travel   
19          19   66246    Male     Loyal Customer   38  Personal Travel   

       Class  Flight Distance  Inflight wifi service  \
0   Eco Plus              460                      3   
1   Business              235                      3   
2   Business             1142                      2   
3   Business              562                      2   
4   Business              214                      3   
5        Eco             1180                      3   
6        Eco             1276                      2   
7   Business             2035                      4   
8   Business              853                      1   
9        Eco             1061                      3   
10       Eco             1182                      4   
11  Eco Plus              308                      2   
12       Eco              834                      1   
13       Eco              946                      4   
14       Eco              453                      3   
15       Eco              486                      2   
16  Business             2123                      3   
17  Business             2075                      4   
18  Business             2486                      4   
19       Eco              460                      2   

    Departure/Arrival time convenient  Ease of Online booking  Gate location  \
0                                   4                       3              1   
1                                   2                       3              3   
2                                   2                       2              2   
3                                   5                       5              5   
4                                   3                       3              3   
5                                   4                       2              1   
6                                   4                       2              3   
7                                   3                       4              4   
8                                   2                       2              2   
9                                   3                       3              4   
10                                  5                       5              4   
11                                  4                       2              2   
12                                  4                       4              4   
13                                  2                       4              3   
14                                  2              

## Exploratory Data Analysis 🔎

In [9]:
df.shape

(103904, 25)

In [10]:
# Convert all column names to lowercase and replace spaces with underscores
df.columns = df.columns.str.lower().str.replace(' ', '_')

In [11]:
df.isna().sum()

unnamed:_0                             0
id                                     0
gender                                 0
customer_type                          0
age                                    0
type_of_travel                         0
class                                  0
flight_distance                        0
inflight_wifi_service                  0
departure/arrival_time_convenient      0
ease_of_online_booking                 0
gate_location                          0
food_and_drink                         0
online_boarding                        0
seat_comfort                           0
inflight_entertainment                 0
on-board_service                       0
leg_room_service                       0
baggage_handling                       0
checkin_service                        0
inflight_service                       0
cleanliness                            0
departure_delay_in_minutes             0
arrival_delay_in_minutes             310
satisfaction    

>The columns unnamed:_0, and 'id' dont feel valuable in the machine learning model.

In [13]:
# Drop unnecessary columns
df = df.drop(columns=['unnamed:_0', 'id'])

> Since there are 103,904 records, removing the 310 rows without arrival_delay_in_minutes, a potentially very important feature in predicting satisfaction, seems to be my best option.

In [15]:
# Drop rows with null values in 'arrival_delay_in_minutes'
df = df.dropna(subset=['arrival_delay_in_minutes'])

df.isna().sum()

gender                               0
customer_type                        0
age                                  0
type_of_travel                       0
class                                0
flight_distance                      0
inflight_wifi_service                0
departure/arrival_time_convenient    0
ease_of_online_booking               0
gate_location                        0
food_and_drink                       0
online_boarding                      0
seat_comfort                         0
inflight_entertainment               0
on-board_service                     0
leg_room_service                     0
baggage_handling                     0
checkin_service                      0
inflight_service                     0
cleanliness                          0
departure_delay_in_minutes           0
arrival_delay_in_minutes             0
satisfaction                         0
dtype: int64

In [16]:
df.dtypes

gender                                object
customer_type                         object
age                                    int64
type_of_travel                        object
class                                 object
flight_distance                        int64
inflight_wifi_service                  int64
departure/arrival_time_convenient      int64
ease_of_online_booking                 int64
gate_location                          int64
food_and_drink                         int64
online_boarding                        int64
seat_comfort                           int64
inflight_entertainment                 int64
on-board_service                       int64
leg_room_service                       int64
baggage_handling                       int64
checkin_service                        int64
inflight_service                       int64
cleanliness                            int64
departure_delay_in_minutes             int64
arrival_delay_in_minutes             float64
satisfacti

> The columns of gender, customer_type, type_of_travel and and class are categorical, and would be more useful to a machine learning model if they were numerical, using pandas get_dummmies funciton will allow us to create a numerical signifier. 

In [18]:
# Encode categorical variables using get_dummies, which converts to 1s and 0s
df = pd.get_dummies(df, columns=['gender', 'customer_type', 'type_of_travel', 'class'], drop_first=True)

In [19]:
df.head(20)

age  flight_distance  inflight_wifi_service  \
0    13              460                      3   
1    25              235                      3   
2    26             1142                      2   
3    25              562                      2   
4    61              214                      3   
5    26             1180                      3   
6    47             1276                      2   
7    52             2035                      4   
8    41              853                      1   
9    20             1061                      3   
10   24             1182                      4   
11   12              308                      2   
12   53              834                      1   
13   33              946                      4   
14   26              453                      3   
15   13              486                      2   
16   26             2123                      3   
17   41             2075                      4   
18   45             2486                      4   
19   38              460                      2   

    departure/arrival_time_convenient  ease_of_online_booking  gate_location  \
0                                   4                       3              1   
1                                   2                       3              3   
2                                   2                       2              2   
3                                   5                       5              5   
4                                   3                       3              3   
5                                   4                       2              1   
6                                   4                       2              3   
7                                   3                       4              4   
8                                   2                       2              2   
9                                   3                       3              4   
10                                  5                       5              4   
11                                  4                       2              2   
12                                  4                       4              4   
13                                  2                       4              3   
14                                  2                       3              2   
15                                  1                       2              3   
16                                  3                       3              3   
17                                  4                       2              4   
18                                  4                       4              4   
19                                  3                       3              2   

    food_and_drink  online_boarding  seat_comfort  inflight_entertainment  \
0                5                3             5                       5   
1                1                3             1                       1   
2                5                5             5                       5   
3                2                2             2                       2   
4                4                5             5                       3   
5                1                2             1                       1   
6                2                2             2                       2   
7                5                5             5                       5   
8                4                3             3                       1   
9                2                3             3                       2   
10               2                5             2                       2   
11               1                2             1                       1   
12               1                1             1                       1   
13               4                4             4                       4   
14               2                3             2                       2   
15               4 

In [20]:
df.dtypes

age                                    int64
flight_distance                        int64
inflight_wifi_service                  int64
departure/arrival_time_convenient      int64
ease_of_online_booking                 int64
gate_location                          int64
food_and_drink                         int64
online_boarding                        int64
seat_comfort                           int64
inflight_entertainment                 int64
on-board_service                       int64
leg_room_service                       int64
baggage_handling                       int64
checkin_service                        int64
inflight_service                       int64
cleanliness                            int64
departure_delay_in_minutes             int64
arrival_delay_in_minutes             float64
satisfaction                          object
gender_Male                             bool
customer_type_disloyal Customer         bool
type_of_travel_Personal Travel          bool
class_Eco 

In [21]:
# Map satisfaction column to numerical values, 0 for neutral or dissatisfied, 1 for satisfied
df['satisfaction'] = df['satisfaction'].map({'neutral or dissatisfied': 0, 'satisfied': 1}).astype(int)

In [22]:
# Count the number of satisfied vs neutral or dissatisfied customers 
df['satisfaction'].value_counts()

satisfaction
0    58697
1    44897
Name: count, dtype: int64

In [23]:
df.head(20)

age  flight_distance  inflight_wifi_service  \
0    13              460                      3   
1    25              235                      3   
2    26             1142                      2   
3    25              562                      2   
4    61              214                      3   
5    26             1180                      3   
6    47             1276                      2   
7    52             2035                      4   
8    41              853                      1   
9    20             1061                      3   
10   24             1182                      4   
11   12              308                      2   
12   53              834                      1   
13   33              946                      4   
14   26              453                      3   
15   13              486                      2   
16   26             2123                      3   
17   41             2075                      4   
18   45             2486                      4   
19   38              460                      2   

    departure/arrival_time_convenient  ease_of_online_booking  gate_location  \
0                                   4                       3              1   
1                                   2                       3              3   
2                                   2                       2              2   
3                                   5                       5              5   
4                                   3                       3              3   
5                                   4                       2              1   
6                                   4                       2              3   
7                                   3                       4              4   
8                                   2                       2              2   
9                                   3                       3              4   
10                                  5                       5              4   
11                                  4                       2              2   
12                                  4                       4              4   
13                                  2                       4              3   
14                                  2                       3              2   
15                                  1                       2              3   
16                                  3                       3              3   
17                                  4                       2              4   
18                                  4                       4              4   
19                                  3                       3              2   

    food_and_drink  online_boarding  seat_comfort  inflight_entertainment  \
0                5                3             5                       5   
1                1                3             1                       1   
2                5                5             5                       5   
3                2                2             2                       2   
4                4                5             5                       3   
5                1                2             1                       1   
6                2                2             2                       2   
7                5                5             5                       5   
8                4                3             3                       1   
9                2                3             3                       2   
10               2                5             2                       2   
11               1                2             1                       1   
12               1                1             1                       1   
13               4                4             4                       4   
14               2                3             2                       2   
15               4 

In [24]:
df.dtypes

age                                    int64
flight_distance                        int64
inflight_wifi_service                  int64
departure/arrival_time_convenient      int64
ease_of_online_booking                 int64
gate_location                          int64
food_and_drink                         int64
online_boarding                        int64
seat_comfort                           int64
inflight_entertainment                 int64
on-board_service                       int64
leg_room_service                       int64
baggage_handling                       int64
checkin_service                        int64
inflight_service                       int64
cleanliness                            int64
departure_delay_in_minutes             int64
arrival_delay_in_minutes             float64
satisfaction                           int32
gender_Male                             bool
customer_type_disloyal Customer         bool
type_of_travel_Personal Travel          bool
class_Eco 

> Earlier, when running get_dummies the created columns turned into Trues and Falses, and not my intended 1s and 0s, Let's fix that.

In [26]:
df[df.select_dtypes(include=['bool']).columns] = df.select_dtypes(include=['bool']).astype(int)


In [27]:
df.dtypes

age                                    int64
flight_distance                        int64
inflight_wifi_service                  int64
departure/arrival_time_convenient      int64
ease_of_online_booking                 int64
gate_location                          int64
food_and_drink                         int64
online_boarding                        int64
seat_comfort                           int64
inflight_entertainment                 int64
on-board_service                       int64
leg_room_service                       int64
baggage_handling                       int64
checkin_service                        int64
inflight_service                       int64
cleanliness                            int64
departure_delay_in_minutes             int64
arrival_delay_in_minutes             float64
satisfaction                           int32
gender_Male                            int32
customer_type_disloyal Customer        int32
type_of_travel_Personal Travel         int32
class_Eco 

In [28]:
df.head(20)

age  flight_distance  inflight_wifi_service  \
0    13              460                      3   
1    25              235                      3   
2    26             1142                      2   
3    25              562                      2   
4    61              214                      3   
5    26             1180                      3   
6    47             1276                      2   
7    52             2035                      4   
8    41              853                      1   
9    20             1061                      3   
10   24             1182                      4   
11   12              308                      2   
12   53              834                      1   
13   33              946                      4   
14   26              453                      3   
15   13              486                      2   
16   26             2123                      3   
17   41             2075                      4   
18   45             2486                      4   
19   38              460                      2   

    departure/arrival_time_convenient  ease_of_online_booking  gate_location  \
0                                   4                       3              1   
1                                   2                       3              3   
2                                   2                       2              2   
3                                   5                       5              5   
4                                   3                       3              3   
5                                   4                       2              1   
6                                   4                       2              3   
7                                   3                       4              4   
8                                   2                       2              2   
9                                   3                       3              4   
10                                  5                       5              4   
11                                  4                       2              2   
12                                  4                       4              4   
13                                  2                       4              3   
14                                  2                       3              2   
15                                  1                       2              3   
16                                  3                       3              3   
17                                  4                       2              4   
18                                  4                       4              4   
19                                  3                       3              2   

    food_and_drink  online_boarding  seat_comfort  inflight_entertainment  \
0                5                3             5                       5   
1                1                3             1                       1   
2                5                5             5                       5   
3                2                2             2                       2   
4                4                5             5                       3   
5                1                2             1                       1   
6                2                2             2                       2   
7                5                5             5                       5   
8                4                3             3                       1   
9                2                3             3                       2   
10               2                5             2                       2   
11               1                2             1                       1   
12               1                1             1                       1   
13               4                4             4                       4   
14               2                3             2                       2   
15               4 

Are there any duplicate rows?

In [30]:
df.duplicated().sum()

0

## Export to CSV 🔽

In [31]:
# Export df to a CSV file
df.to_csv('data/train_cleaned.csv', index=False)